framework/src/utils/SlepcSupport.C

```diff
@@ -356,9 +433,17 @@ moosePetscSNESFormFunction(SNES /*snes*/, Vec x, Vec r, void * ctx, Moose::Kerne
   X_global.localize(*sys.current_local_solution.get());
 
   R.zero();
-
-  eigen_problem->computeResidualType(*sys.current_local_solution.get(), R, type);
-
+  if (type == Moose::KT_EIGEN && eigen_problem->residualInitialed())
+  {
+    eigen_problem->computeResidualTypeBx(*sys.current_local_solution.get(), R, type);
+    eigen_problem->setResidualInitialed(false);
+  }
+  else
+  {
+    eigen_problem->computeResidualType(*sys.current_local_solution.get(), R, type);
+    if (type == Moose::KT_NONEIGEN)
+      eigen_problem->setResidualInitialed(true);
```

Comment:
What's odd about it is the fact that it's an internal state variable that's wide open to be set by anyone. I don't understand why you can't set the internal state from within the object when you compute residuals based on all of the parameters that you pass in when computing those residuals. To me, this variable should be exposed as a setter from an encapsulation point of view.